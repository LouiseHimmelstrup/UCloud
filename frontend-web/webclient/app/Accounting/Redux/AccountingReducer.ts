import { Reducer as ReduxReducer } from "redux";
import { Type as ActionType, Tag } from "./AccountingActions";
import { Type as ReduxType, init, ResourceState } from "./AccountingObject";
import { loadableEventToContent } from "LoadableContent";
import { emptyResourceState } from "./AccountingObject";

export interface Reducer {
    accounting: ReduxReducer<ReduxType>
}

function genericReduce(state: ReduxType, resource: string, newPartialState: Partial<ResourceState>): ReduxType {
    const result = JSON.parse(JSON.stringify(state));
    const currentState: ResourceState = result.resources[resource] || emptyResourceState();
    const newState: ResourceState = { ...currentState, ...newPartialState };
    result.resources[resource] = newState;
    return result;
}

const reducer = (state: ReduxType = init().accounting, action: ActionType): ReduxType => {
    switch (action.type) {
        case Tag.RECEIVE_CHART: {
            return genericReduce(state, action.payload.resource,
                { chart: loadableEventToContent(action.payload.event) });
        }

        case Tag.RECEIVE_EVENTS: {
            return genericReduce(state, action.payload.resource,
                { events: loadableEventToContent(action.payload.event) });
        }

        case Tag.RECEIVE_USAGE: {
            return genericReduce(state, action.payload.resource,
                { usage: loadableEventToContent(action.payload.event) });
        }

        default: {
            return state;
        }
    }
};

export default reducer;