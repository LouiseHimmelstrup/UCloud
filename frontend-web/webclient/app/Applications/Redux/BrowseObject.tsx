import { LoadableContent, emptyLoadableContent } from "LoadableContent";
import { Page } from "Types";
import { ApplicationMetadata, WithAppMetadata } from "Applications";

export interface Type {
    applications: Map<string, Page<WithAppMetadata>>
    loading: boolean
}

export interface Wrapper {
    applicationsBrowse: Type
}

export const init = (): Wrapper => ({
    applicationsBrowse: {
        applications: new Map(),
        loading: false
    }
});