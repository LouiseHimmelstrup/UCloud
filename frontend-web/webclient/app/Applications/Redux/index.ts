import * as ViewObject from "./ViewObject";
import * as ViewReducer from "./ViewReducer";
import * as ApplicationObject from "./ApplicationsObject";
import * as ApplicationReducer from "./ApplicationsReducer";

export type Reducers = ViewReducer.Reducer & ApplicationReducer.Reducer;
export type Objects = ViewObject.Wrapper & ApplicationObject.Wrapper;

export function init(): Objects {
    return { 
        ...ViewObject.init(),
        ...ApplicationObject.init()
    }
}

export const reducers: Reducers = {
    applicationView: ViewReducer.default,
    applicationsBrowse: ApplicationReducer.default
};