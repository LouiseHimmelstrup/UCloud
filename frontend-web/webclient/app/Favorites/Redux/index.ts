import * as FavoritesObject from "./FavoriteObject";
import * as FavoritesReducer from "./FavoritesReducer";

export type Reducers = FavoritesReducer.Reducer;
export type Objects = FavoritesObject.Wrapper;

export function init(): Objects {
    return {
        ...FavoritesObject.init()
    }
}

export const reducers: Reducers = {
    favorites: FavoritesReducer.default
}