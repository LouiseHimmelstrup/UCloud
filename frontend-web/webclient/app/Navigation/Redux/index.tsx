import {PRODUCT_NAME} from "../../../site.config.json";
import {SetLoadingAction} from "@/Types";
import {useDispatch} from "react-redux";
import {useEffect} from "react";
import {initStatus, StatusReduxObject} from "@/DefaultObjects";

export type Index = UpdatePageTitleAction | SetLoading;

export type UpdatePageTitleAction = PayloadAction<typeof UPDATE_PAGE_TITLE, {title: string}>;
/**
 * Sets the title of the window. Stores in the redux store as well
 * @param {string} title the title to be set
 */
export const updatePageTitle = (title: string): UpdatePageTitleAction => ({
    type: UPDATE_PAGE_TITLE,
    payload: {title}
});

type SetLoading = SetLoadingAction<typeof SET_STATUS_LOADING>;
export const setLoading = (loading: boolean): SetLoading => ({
    type: SET_STATUS_LOADING,
    payload: {loading}
});

export interface SetStatusLoading {
    setLoading: (loading: boolean) => void;
}

export function useTitle(title: string): void {
    const dispatch = useDispatch();
    useEffect(() => {
        dispatch(updatePageTitle(title));
        return () => {
            dispatch(updatePageTitle(""));
        };
    }, [title]);
}

export function useLoading(loading: boolean): void {
    const dispatch = useDispatch();
    useEffect(() => {
        dispatch(setLoading(loading));
    }, [loading]);
}

export const UPDATE_PAGE_TITLE = "UPDATE_PAGE_TITLE";
export const SET_STATUS_LOADING = "SET_STATUS_LOADING";

export const statusReducer = (state: StatusReduxObject = initStatus(), action: Index): StatusReduxObject => {
    switch (action.type) {
        case UPDATE_PAGE_TITLE:
            document.title = `${PRODUCT_NAME} | ${action.payload.title}`;
            return {...state, ...action.payload};
        case SET_STATUS_LOADING:
            return {...state, ...action.payload};
        default: {
            return state;
        }
    }
};

