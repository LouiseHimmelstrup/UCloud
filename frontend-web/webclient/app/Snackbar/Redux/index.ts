import snackbar from "./SnackbarsReducer";
import { Snack } from "Snackbar/Snackbars";

export const init = (): Wrapper => ({
    snackbar: {
        snackbar: [] as Snack[],
        nextId: 0
    }
});

export const reducers = {
    snackbar
};

export interface Wrapper {
    snackbar: Object
}

export interface Object {
    snackbar: Snack[]
    nextId: number
}