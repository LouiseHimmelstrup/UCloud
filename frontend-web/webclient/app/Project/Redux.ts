import {File} from "Files";
import {Action} from "redux";
import {Dictionary, Page} from "Types";
import {getParentPath, resolvePath} from "Utilities/FileUtilities";

export interface State {
    project?: string;
    files?: Dictionary<ProjectFilesCacheEntry>;
}

export interface ProjectFilesCacheEntry {
    project: string;
    path: string;
    files: Page<File>;
}

export const initialState = {project: getStoredProject() ?? undefined};

interface SetProjectAction extends Action<"SET_PROJECT"> {
    project?: string;
}

export function setProjectAction(project?: string): SetProjectAction {
    return {project, type: "SET_PROJECT"};
}

interface UpdateProjectFileAction extends Action<"UPDATE_PROJECT_FILE"> {
    newFile: File;
}

export function updateProjectFileAction(newFile: File): UpdateProjectFileAction {
    return {newFile, type: "UPDATE_PROJECT_FILE"};
}

type CacheProjectFilesAction = Action<"CACHE_PROJECT_FILES"> & ProjectFilesCacheEntry;

export function cacheProjectFilesAction(entry: ProjectFilesCacheEntry): CacheProjectFilesAction {
    return {...entry, type: "CACHE_PROJECT_FILES"};
}

type ClearDirtFlagAction = Action<"CACHE_PROJECT_FILES_CLEAR_DIRTY">;

export function clearProjectFilesDirtyFlag(): ClearDirtFlagAction {
    return {type: "CACHE_PROJECT_FILES_CLEAR_DIRTY"};
}

type ProjectAction = SetProjectAction | CacheProjectFilesAction | UpdateProjectFileAction | ClearDirtFlagAction;

export const reducer = (state: State = initialState, action: ProjectAction) => {
    switch (action.type) {
        case "SET_PROJECT": {
            setStoredProject(action.project ?? null);
            return {...state, project: action.project};
        }

        case "CACHE_PROJECT_FILES": {
            const existingDirectoryOrInitial = state.files ?? {};
            const files = {...existingDirectoryOrInitial};
            files[action.path] = action;
            return {...state, files};
        }

        case "UPDATE_PROJECT_FILE": {
            const parentPath = resolvePath(getParentPath(action.newFile.path));
            const newFile = {...action.newFile, frontend: {dirty: true}};
            const cachedParent: ProjectFilesCacheEntry = state.files?.[parentPath] ??
                {
                    project: state.project!,
                    path: parentPath,
                    files: {
                        items: [newFile],
                        pageNumber: 0,
                        itemsPerPage: 50,
                        itemsInTotal: 1,
                        pagesInTotal: 1
                    }
                };

            const directoryCopy = cachedParent.files.items.slice();
            for (let i = 0; i < directoryCopy.length; i++) {
                const file = directoryCopy[i];
                if (file.path === newFile.path) {
                    directoryCopy[i] = newFile;
                    break;
                }
            }

            const files = {...state.files};
            files[parentPath] = {...cachedParent, files: {...cachedParent.files, items: directoryCopy}};
            return {...state, files};
        }

        case "CACHE_PROJECT_FILES_CLEAR_DIRTY": {
            const files: Dictionary<ProjectFilesCacheEntry> = {...state.files};
            Object.keys(files).forEach(key => {
                const current = files[key];
                files[key] = {
                    ...current,
                    files: {
                        ...current.files,
                        items: current.files.items.map(it => ({...it, frontend: undefined}))
                    }
                };
            });

            return {...state, files};
        }

        default:
            return state;
    }
};

function getStoredProject(): string | null {
    return window.localStorage.getItem("project") ?? null;
}

function setStoredProject(value: string | null) {
    if (value === null) {
        window.localStorage.removeItem("project");
    } else {
        window.localStorage.setItem("project", value);
    }
}

