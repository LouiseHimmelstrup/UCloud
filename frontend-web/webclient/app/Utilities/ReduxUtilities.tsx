import dashboard from "Dashboard/Redux/DashboardReducer";
import {initObject} from "DefaultObjects";
import header, {CONTEXT_SWITCH, USER_LOGIN, USER_LOGOUT} from "Navigation/Redux/HeaderReducer";
import sidebar from "Navigation/Redux/SidebarReducer";
import status from "Navigation/Redux/StatusReducer";
import notifications from "Notifications/Redux/NotificationsReducer";
import * as ProjectRedux from "Project/Redux";
import {Action, AnyAction, combineReducers, createStore, Store} from "redux";
import {composeWithDevTools} from "redux-devtools-extension";
import {createResponsiveStateReducer, responsiveStoreEnhancer} from "redux-responsive";
import simpleSearch from "Search/Redux/SearchReducer";
import {responsiveBP} from "ui-components/theme";
import avatar from "UserSettings/Redux/AvataaarReducer";
import hookStore from "Utilities/ReduxHooks";

export function configureStore(
    initialObject: ReduxObject,
    reducers,
    enhancers?
): Store<ReduxObject, AnyAction> {
    const combinedReducers = combineReducers<ReduxObject, AnyAction>(reducers);
    const rootReducer = (state: ReduxObject, action: Action): ReduxObject => {
        if ([USER_LOGIN, USER_LOGOUT, CONTEXT_SWITCH].some(it => it === action.type)) {
            state = initObject();
        }
        return combinedReducers(state, action);
    };
    return createStore<ReduxObject, AnyAction, {}, {}>(rootReducer, initialObject, composeWithDevTools(enhancers));
}

export const responsive = createResponsiveStateReducer(
    responsiveBP,
    {infinity: "xxl"}
);

export const store = configureStore(initObject(), {
    dashboard,
    header,
    status,
    sidebar,
    notifications,
    simpleSearch,
    hookStore,
    avatar,
    loading,
    project: ProjectRedux.reducer,
    responsive: createResponsiveStateReducer(
        responsiveBP,
        {infinity: "xxl"}),
}, responsiveStoreEnhancer);

function loading(state = false, action: {type: string}): boolean {
    switch (action.type) {
        case "LOADING_START":
            return true;
        case "LOADING_END":
            return false;
        default:
            return state;
    }
}
