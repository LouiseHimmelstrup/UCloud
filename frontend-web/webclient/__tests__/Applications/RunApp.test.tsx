import * as React from "react";
import { create } from "react-test-renderer";
import { configureStore } from "Utilities/ReduxUtilities";
import { init } from "Applications/Redux/BrowseObject";
import applications from "Applications/Redux/BrowseReducer";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router";
import "jest-styled-components";

describe("RunApp component", () => {

    // FIXME Requires match props, but for some reason isn't allowed
    test.skip("Mount", () => {
        const store = configureStore({ applications: init() }, { applications })
        expect(create(
            <Provider store={store}>
                <MemoryRouter>
                    {/* <RunApp match={{ params: { appName: "appName", appVersion: "appVersion" } }} /> */}
                </MemoryRouter>
            </Provider>).toJSON()
        ).toMatchSnapshot();
    });

});
