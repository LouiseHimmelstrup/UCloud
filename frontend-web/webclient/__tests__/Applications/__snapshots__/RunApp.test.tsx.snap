// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RunApp component Mount 1`] = `
<div
  className="ui container sixteen column grid"
>
  <div
    className="sixteen wide column"
  >
    <i
      aria-hidden="true"
      className="circle notched loading icon"
    />
    <h1
      className="ui header"
    >
      <div
        className="content float-right"
      >
        <button
          className="ui button"
          onClick={[Function]}
          role="button"
        >
          Export parameters
        </button>
        <label
          className="ui button"
          onClick={[Function]}
          role="button"
        >
          Import parameters
          <input
            className="import-parameters"
            onChange={[Function]}
            type="file"
          />
        </label>
        <a
          className="ui blue basic button"
          href="/appDetails/appName/appVersion/"
          onClick={[Function]}
          role="button"
        >
          More information
        </a>
      </div>
      <div
        className="content"
      >
        appName
        <span
          className="app-favorite-padding"
        >
          <div
            className="ui star huge rating"
            onClick={[Function]}
            onMouseLeave={[Function]}
            role="radiogroup"
          >
            <i
              aria-checked={false}
              aria-posinset={1}
              aria-setsize={1}
              className="icon"
              onClick={[Function]}
              onKeyUp={[Function]}
              onMouseEnter={[Function]}
              role="radio"
              tabIndex={0}
            />
          </div>
        </span>
        <h4>
          appVersion
        </h4>
        <h4>
          Author
          : 
          
        </h4>
      </div>
    </h1>
    <form
      className="ui form"
      onSubmit={[Function]}
    >
      <div
        className="disabled field"
      >
        <label>
          Jobname
        </label>
        <div
          className="ui disabled input"
        >
          <input
            disabled={true}
            onChange={[Function]}
            placeholder="Jobname will be assigned if field left empty"
            tabIndex={-1}
            type="text"
          />
        </div>
      </div>
      <div
        className="disabled field"
      >
        <label>
          Tags (Separated by space)
        </label>
        <div
          className="ui disabled input"
        >
          <input
            disabled={true}
            onChange={[Function]}
            placeholder="Assign tags to jobs"
            tabIndex={-1}
            type="text"
          />
        </div>
      </div>
      <div
        className="disabled field"
      >
        <label>
          Comment
        </label>
        <textarea
          disabled={true}
          onChange={[Function]}
          onInput={[Function]}
          placeholder="Comment..."
          rows={3}
          style={
            Object {
              "resize": "",
            }
          }
        />
      </div>
      <button
        className="ui blue button"
        onClick={[Function]}
        role="button"
      >
        Submit
      </button>
    </form>
  </div>
</div>
`;
